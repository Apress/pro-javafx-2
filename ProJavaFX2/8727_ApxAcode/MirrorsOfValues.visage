/*
 * @test
 * @run
 */

import java.lang.Math.*;
import visage.reflect.*;

public class Point {
  public var x: Number;
  public var y: Number;

  public function dist() {
    sqrt(x * x + y * y);
  }

  public override function toString() {
    "Point \{ x: {x}, y: {y} \}"
  }
}

public function run() {
  var i = 1024;
  var o = Point { x: 3, y: 4 };
  var seq = [3.14159, 2.71828];
  var func = o.dist;

  var context: VisageLocal.Context = VisageLocal.getContext();

  println("Examining...");
  var mirrorOfI = context.mirrorOf(i);
  print("  original: "); println({i});
  print("  mirror: "); println(mirrorOfI);
  print("  type: "); println({mirrorOfI.getType()});
  print("  back from mirror: "); println({mirrorOfI.asObject()});

  println("Examining...");
  var mirrorOfO = context.mirrorOf(o);
  print("  original: "); println({o});
  print("  mirror: "); println(mirrorOfO);
  print("  type: "); println({mirrorOfO.getType()});
  print("  back from mirror: "); println({mirrorOfO.asObject()});

  println("Exemining...");
  var seqType = context.getNumberType().getSequenceType();
  var mirrorOfSeq = context.mirrorOf(seq, seqType);
  print("  original: "); println({seq});
  print("  mirror: "); println(mirrorOfSeq);
  print("  type: "); println({mirrorOfSeq.getType()});
  print("  back from mirror: "); println({mirrorOfSeq.asObject()});

  println("Exemining...");
  var classType = context.findClass("MirrorsOfValues.Point");
  var funcMember = classType.getFunction("dist");
  var funcType = funcMember.getType();
  var mirrorOfFunc = context.mirrorOf(func, funcType);
  print("  original: "); println({func});
  print("  mirror: "); println(mirrorOfFunc);
  print("  type: "); println({mirrorOfFunc.getType()});
  print("  back from mirror: "); println({mirrorOfFunc.asObject()});
}
