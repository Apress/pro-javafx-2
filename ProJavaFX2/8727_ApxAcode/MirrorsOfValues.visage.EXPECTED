Examining...
  original: 1024
  mirror: IntegerValue(1024)
  type: Integer
  back from mirror: 1024
Examining...
  original: Point { x: 3.0, y: 4.0 }
// @Skip  mirror: visage.reflect.VisageLocal$ObjectValue@37a786c3
  type: class MirrorsOfValues.Point
  back from mirror: Point { x: 3.0, y: 4.0 }
Exemining...
  original: [ 3.14159, 2.71828 ]
// @Skip  mirror: visage.reflect.VisageLocal$SequenceValue@b32e13d
  type: Float[]
  back from mirror: [ 3.14159, 2.71828 ]
Exemining...
  original: MirrorsOfValues$1Local$1.function<0>
// @Skip  mirror: visage.reflect.VisageLocal$FunctionValue@182d9c06
  type: function():Double
  back from mirror: MirrorsOfValues$1Local$1.function<0>
